tests/functional_tests/ffront_tests/test_program.py

```diff
@@ -207,3 +206,25 @@ def prog(
 
     assert np.allclose(a, out_a)
     assert np.allclose(b, out_b)
+
+
+def test_wrong_argument_type(fieldview_backend, copy_program_def):
+    size = 10
+    inp = np_as_located_field(JDim)(np.ones((size,)))
+    out = np_as_located_field(JDim)(np.zeros((size,)))
+
+    copy_program = program(copy_program_def, backend=fieldview_backend)
+
+    with pytest.raises(ProgramTypeError) as exc_info:
+        # program is defined on Field[[IDim], ...], but we call with
+        #  Field[[JDim], ...]
+        copy_program(inp, out, offset_provider={})
+
+    msgs = [
+        "- Expected 0-th argument to be of type Field\[\[IDim], dtype=float64\],"
+        " but got Field\[\[JDim\], dtype=float64\].",
+        "- Expected 1-th argument to be of type Field\[\[IDim], dtype=float64\],"
+        " but got Field\[\[JDim\], dtype=float64\].",
+    ]
+    for msg in msgs:
+        assert re.search(msg, exc_info.value.__context__.args[0]) is not None
```

Comment:
more style comment: if you need a complicated test condition, it's an indication that you are testing too much in this test. I'd just make one argument wrong and use standard pytest.raises(..., match=r"...") mechanism.

Reply:
That's true. I had chosen to use the wrong dimensions for both args as the `out` argument is special. But only on field operator not program level so this is not needed here. I only test for one argument now.

The purpose of this test is to test that the user becomes a meaningful error message signaling what is wrong. The outer exception that `pytest.raises(..., match=r"...")` matches on, i.e. `invalid call to program ...`, does not contain that information. Hence I am testing the inner exception, which I inconveniently have to access using `exc_info.value.__context__`. The `raise outer_exception from inner_exception` pattern we use is anyway not suitable for giving users good error messages and we should change that. Then this will disappear and we just use `pytest.raises(..., match=r"...")`.

Reply:
ah, I missed that...

Reply:
Just a comment here for future reference/discussions: when using the `raise from` pattern, the inner exception is easily reachable through the `__cause__`  attribute (see [PEP-3134](https://peps.python.org/pep-3134/)). Also note that using `raise from` pattern does not prevent to give good error messages since the message of the outer exception can always be customized on creation when catching the inner exception.